### PR TITLE
[BEAM-8372] Job server submitting UberJars directly to Flink Runner.

### PR DESCRIPTION
**Pull request by robertwb**
One can now use the FlinkRunner with the --flink_master_url=flink-master-url
and an UberJar will automatically be created with the artifacts and pipeline
definition, uploaded to the master, and started.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by robertwb** [comment]
CC: @mxm

File: sdks/python/apache_beam/runners/portability/abstract_job_service.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  """
+  def __init__(self):
+    self._jobs = {}
+
+  def create_beam_job(self, preparation_id, job_name, pipeline, options):
+    """Returns an instance of AbstractBeamJob specific to this servicer."""
+    raise NotImplementedError(type(self))
+
+  def Prepare(self, request, context=None, timeout=None):
+    # For now, just use the job name as the job id.
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/abstract_job_service.py]: Strictly speaking, the job name would just be a prefix of the job id.

File: sdks/python/apache_beam/runners/portability/artifact_service.py [hunk 2]
```diff
@@ -23,51 +23,61 @@
...
 
-class BeamFilesystemArtifactService(
+class AbstractArtifactService(
     beam_artifact_api_pb2_grpc.ArtifactStagingServiceServicer,
     beam_artifact_api_pb2_grpc.ArtifactRetrievalServiceServicer):
 
   _DEFAULT_CHUNK_SIZE = 2 << 20  # 2mb
 
-  def __init__(self, root, chunk_size=_DEFAULT_CHUNK_SIZE):
+  def __init__(self, root, chunk_size=None):
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/artifact_service.py]: I don't see the difference here.

File: sdks/python/apache_beam/runners/portability/flink_runner.py [hunk 3]
```diff
@@ -29,7 +32,13 @@
 
 class FlinkRunner(portable_runner.PortableRunner):
   def default_job_server(self, options):
-    return job_server.StopOnExitJobServer(FlinkJarJobServer(options))
+    flink_master_url = options.view_as(FlinkRunnerOptions).flink_master_url
+    flink_master_url = 'http://localhost:8081'
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/flink_runner.py]: This looks wrong.

File: sdks/python/apache_beam/runners/portability/flink_runner.py [hunk 4]
```diff
@@ -29,7 +32,13 @@
...
+    if flink_master_url == '[local]' or sys.version_info < (3, 6):
+      # TOOD: Also default to LOOPBACK?
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/flink_runner.py]: Seems reasonable. Maybe file a JIRA for it to link here.

File: sdks/python/apache_beam/runners/portability/artifact_service.py [hunk 5]
```diff
@@ -77,47 +87,120 @@ def PutArtifact(self, request_iterator, context=None):
...
+          chunk = True
+          while chunk:
+            chunk = fin.read(self._chunk_size)
+            yield beam_artifact_api_pb2.ArtifactChunk(data=chunk)
+        break
+    else:
+      raise ValueError('Unknown artifact: %s' % request.name)
+
+
+class ZipFileArtifactService(AbstractArtifactService):
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/artifact_service.py]: Can you add a docstring?

Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/artifact_service.py]: If this is the piece that depends on Python 3.6, could you document that here? And maybe add a check if you think it's needed?

File: sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        self.executable_jar(),
+        job_id,
+        job_name,
+        pipeline,
+        options)
+
+
+class FlinkBeamJob(abstract_job_service.AbstractBeamJob):
+
+  # These must agree with those defined in PortablePipelineJarUtils.java.
```
[the remark this answers is not in the document]
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py]: s/open/merged so you'll have to rebase, sorry about that

[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py]: Thanks for the heads up.

File: sdks/python/apache_beam/runners/portability/artifact_service.py (hunk 2 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/artifact_service.py]: Missed that, makes sense.

**Comment by robertwb** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/runners/portability/flink_runner.py [hunk 7]
```diff
@@ -29,7 +32,12 @@
 
 class FlinkRunner(portable_runner.PortableRunner):
   def default_job_server(self, options):
-    return job_server.StopOnExitJobServer(FlinkJarJobServer(options))
+    flink_master_url = options.view_as(FlinkRunnerOptions).flink_master_url
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/flink_runner.py]: This should be `flink_master` to be compliant with https://github.com/apache/beam/blob/14b5ef8b3e63d77f5f34730ce10b9ed581380ffa/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkPipelineOptions.java#L66

**Comment by mxm** [comment]
Thanks @robertwb. This looks great. Having a look now.

File: sdks/python/apache_beam/runners/portability/artifact_service.py [hunk 8]
```diff
@@ -77,47 +88,131 @@ def PutArtifact(self, request_iterator, context=None):
...
+  submitting to an upstream runner's cluster.
+
+  Writing to zip files requires Python 3.6+.
+  """
+
+  def __init__(self, path, chunk_size=None):
+    if sys.version_info < (3, 6):
+      raise RuntimeError(
+          'Writing to zip files requires Python 3.6+, '
+          'but current version is %s' % sys.version)
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/artifact_service.py]: That is a bit unfortunate for portability, but ok in terms of code simplicity.

File: sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+
+from apache_beam.portability.api import beam_artifact_api_pb2_grpc
+from apache_beam.portability.api import beam_job_api_pb2
+from apache_beam.portability.api import endpoints_pb2
+from apache_beam.runners.portability import abstract_job_service
+from apache_beam.runners.portability import artifact_service
+from apache_beam.runners.portability import job_server
+
+
+class FlinkUberJarJobServer(abstract_job_service.AbstractJobServiceServicer):
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py]: Would be nice to have a docstring here.
```suggestion
class FlinkUberJarJobServer(abstract_job_service.AbstractJobServiceServicer):
"""A Job server which submits a self-contained Jar to a Flink cluster. 
The jar contains the Beam dependencies and the Python artifacts."""
```

File: sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    return FlinkBeamJob(
+        self._master_url,
+        self.executable_jar(),
+        job_id,
+        job_name,
+        pipeline,
+        options)
+
+
+class FlinkBeamJob(abstract_job_service.AbstractBeamJob):
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py]: Would be nice to have a docstring here.
```suggestion
class FlinkBeamJob(abstract_job_service.AbstractBeamJob):
"""Runs a Beam job on Flink by staging all contents into a Jar 
and uploading it via the Flink Rest API."""
```

File: sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        'FAILED': beam_job_api_pb2.JobState.FAILED,
+        'CANCELLING': beam_job_api_pb2.JobState.RUNNING,
+        'CANCELED': beam_job_api_pb2.JobState.CANCELLED,
+        'FINISHED': beam_job_api_pb2.JobState.DONE,
+        'RESTARTING': beam_job_api_pb2.JobState.RUNNING,
+        'SUSPENDED': beam_job_api_pb2.JobState.RUNNING,
+        'RECONCILING': beam_job_api_pb2.JobState.RUNNING,
+        'IN_PROGRESS': beam_job_api_pb2.JobState.RUNNING,
+        'COMPLETED': beam_job_api_pb2.JobState.DONE,
+    }.get(flink_status, beam_job_api_pb2.JobState.DONE)
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/flink_uber_jar_job_server.py]: Shouldn't this return `JobState.UNSPECIFIED` by default?

File: sdks/python/apache_beam/runners/portability/flink_runner.py (hunk 7 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/flink_runner.py]: Sorry, I merged before refreshing and seeing your comments. https://github.com/apache/beam/pull/9803

File: sdks/python/apache_beam/runners/portability/artifact_service.py (hunk 8 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/artifact_service.py]: Yeah. 3.5 is pretty old though, and 2.7 is EOL, so not worth implementing our own backport.

**Comment by udim** [comment]
I opened a bug for flakiness in test_concurrent_requests. Perhaps this PR caused it?
https://issues.apache.org/jira/browse/BEAM-8416

**Comment by tvalentyn** [comment]
+1, also hit by the flakes (https://issues.apache.org/jira/browse/BEAM-8426), didn't notice that the bug was already open, thanks, @udim.

**Pull request merged**